File: src/dotnet/cdk/Services/Analytics/AnalyticsService.cs
using System.Collections.Generic;
using Amazon.CDK;
using Amazon.CDK.AWS.Events;
using Amazon.CDK.AWS.Events.Targets;
using Amazon.CDK.AWS.Lambda.EventSources;
using Amazon.CDK.AWS.SecretsManager;
using Constructs;
using ServerlessGettingStarted.CDK.Constructs;

namespace ServerlessGettingStarted.CDK.Services.Analytics;

public record AnalyticsServiceProps(SharedProps Shared, ISecret DdApiKeySecret, IEventBus SharedEventBus);

public class AnalyticsService : Construct
{
    public AnalyticsService(Construct scope, string id, AnalyticsServiceProps props) : base(scope, id)
    {
        var analyticsQueue = new ResilientQueue(this, "AnalyticsQueue",
            new ResilientQueueProps($"{props.Shared.ServiceName}-Events", props.Shared.Env));
        
        var analyticsHandlerFunction = new InstrumentedFunction(this, "AnalyticsHandlerFunction",
            new FunctionProps(props.Shared,"AnalyticsHandlerFunction", "../src/Analytics/Analytics.Adapters/",
                "Analytics.Adapters::Analytics.Adapters.HandlerFunctions_HandleEvents_Generated::HandleEvents", new Dictionary<string, string>(), props.DdApiKeySecret));
        analyticsHandlerFunction.Function.AddEventSource(new SqsEventSource(analyticsQueue.Queue, new SqsEventSourceProps()
        {
            ReportBatchItemFailures = true,
            BatchSize = 10,
            MaxBatchingWindow = Duration.Seconds(10)
        }));

        var eventCatchAllRule = new Rule(this, "AnalyticsEventCatchAllRule", new RuleProps()
        {
            EventBus = props.SharedEventBus,
        });
        eventCatchAllRule.AddEventPattern(new EventPattern()
        {
            Source = Match.Prefix(props.Shared.Env)
        });
        eventCatchAllRule.AddTarget(new SqsQueue(analyticsQueue.Queue));
    }
}

File: src/go/cdk/analytics-service/analyticsService.go
package analyticsservice

import (
	sharedprops "cdk/shared"
	sharedconstructs "cdk/sharedConstructs"

	"github.com/aws/aws-cdk-go/awscdk/v2/awsevents"
	"github.com/aws/aws-cdk-go/awscdk/v2/awseventstargets"
	"github.com/aws/aws-cdk-go/awscdk/v2/awslambdaeventsources"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type AnalyticsServiceProps struct {
	SharedProps    sharedprops.SharedProps
	SharedEventBus awsevents.IEventBus
}

func NewAnalyticsService(scope constructs.Construct, id string, props *AnalyticsServiceProps) {
	analyticsCatchAllEventsQueue := sharedconstructs.NewResiliantQueue(scope, "GoAnalyticsQueue", &sharedconstructs.ResiliantQueueProps{
		SharedProps: props.SharedProps,
		QueueName:   "GoAnalyticsQueue",
	})

	environmentVariables := make(map[string]*string)

	eventAnalyticsHandler := sharedconstructs.NewInstrumentedFunction(scope, "EventAnalytics", &sharedconstructs.InstrumentedFunctionProps{
		SharedProps:          props.SharedProps,
		FunctionName:         "GoEventAnalytics",
		Entry:                "../src/analytics-service/handle-events/",
		EnvironmentVariables: environmentVariables,
	})

	eventAnalyticsHandler.Function.AddEventSource(awslambdaeventsources.NewSqsEventSource(analyticsCatchAllEventsQueue.Queue, &awslambdaeventsources.SqsEventSourceProps{
		ReportBatchItemFailures: jsii.Bool(true),
	}))

	productCreatedRule := awsevents.NewRule(scope, jsii.String("Analytics-CatchAll"), &awsevents.RuleProps{
		EventBus: props.SharedEventBus,
	})

	productCreatedRule.AddEventPattern(&awsevents.EventPattern{
		Source: awsevents.Match_Prefix(&props.SharedProps.Env),
	})
	productCreatedRule.AddTarget(awseventstargets.NewSqsQueue(analyticsCatchAllEventsQueue.Queue, &awseventstargets.SqsQueueProps{}))
}
